sphinxcontrib/datatemplates/directive.py

```diff
@@ -26,39 +27,67 @@ class DataTemplate(rst.Directive):
+        try:
+            if encoding is None:
+                f = open(filename, 'r', newline='')
+            else:
+                f = codecs.open(filename, 'r', newline='', encoding=encoding)
+            dialect = self.options.get('csvdialect')
+            if dialect == "auto":
+                sample = f.read(8192)
+                f.seek(0)
+                sniffer = csv.Sniffer()
+                dialect = sniffer.sniff(sample)
+            if 'csvheaders' in self.options:
+                if dialect is None:
+                    r = csv.DictReader(f)
+                else:
+                    r = csv.DictReader(f, dialect=dialect)
+            else:
+                if dialect is None:
+                    r = csv.reader(f)
+                else:
+                    r = csv.reader(f, dialect=dialect)
+            return list(r)
+        finally:
+            f.close()
+
+    def _load_json(self, filename, encoding):
+        try:
+            if encoding is None:
+                f = open(filename, 'r')
+            else:
+                f = codecs.open(filename, 'r', encoding=encoding)
+            return json.load(f)
+        finally:
+            f.close()
+
+    def _load_yaml(self, filename, encoding):
+        try:
+            if encoding is None:
+                f = open(filename, 'r')
+            else:
+                f = codecs.open(filename, 'r', encoding=encoding)
+            return yaml.safe_load(f)
+        finally:
+            f.close()
+
+    def _load_data(self, env, data_source, encoding):
         rel_filename, filename = env.relfn2path(data_source)
         if data_source.endswith('.yaml'):
-            with open(filename, 'r') as f:
-                return yaml.safe_load(f)
+            return self._load_yaml(filename, encoding)
         elif data_source.endswith('.json'):
-            with open(filename, 'r') as f:
-                return json.load(f)
+            return self._load_json(filename, encoding)
         elif data_source.endswith('.csv'):
-            with open(filename, 'r', newline='') as f:
-                dialect = self.options.get('csvdialect')
-                if dialect == "auto":
-                    sample = f.read(8192)
-                    f.seek(0)
-                    sniffer = csv.Sniffer()
-                    dialect = sniffer.sniff(sample)
-                if 'csvheaders' in self.options:
-                    if dialect is None:
-                        r = csv.DictReader(f)
-                    else:
-                        r = csv.DictReader(f, dialect=dialect)
-                else:
-                    if dialect is None:
-                        r = csv.reader(f)
-                    else:
-                        r = csv.reader(f, dialect=dialect)
-                return list(r)
+            return self._load_csv(filename, encoding)
         elif "xml" in mimetypes.guess_type(data_source)[0]:
+            # there are many XML based formats
```

Comment:
Although I notice there isn't a separate local load function for XML. Does the XML parser deal with encoding itself?

Reply:
Exactly. XML files state their encoding at their beginning or must have it autodetectable (See https://www.w3.org/TR/2006/REC-xml11-20060816/#charencoding ).
It is not actually stated in the docs but [xml.etree.ElementTree.parse](https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.parse) seems to do autodetection.

Reply:
Yeah, that's what I thought I remembered. It has been a while since I've done any serious XML.